File: 02_Algorithm/algorithm_python/Algorithm_Session/Week4/boj_2667.py
import sys
from collections import deque

input = sys.stdin.readline

n = int(input())

arr = []
rows = [-1, 1, 0, 0]
cols = [0, 0, -1, 1]
size = len(arr)
ans = []

for _ in range(n):
    arr.append(list(map(int, input().rstrip())))

def bfs(i, j):
    q = deque()
    q.append((i, j))
    arr[i][j] = 0
    cnt = 0

    while q:
        curR, curC = q.popleft()
        cnt += 1

        print(f"curR: {curR}, curC: {curC}")

        for i in range(4):
            nextR = curR + rows[i]
            nextC = curC + cols[i]

            if not(nextR < 0 or nextC < 0 or nextR >= size or nextC >= size or arr[nextR][nextC] == 0):
                q.append((nextR, nextC))
                arr[nextR][nextC] = 0
    
    ans.append(cnt)

for i in range(n):
    for j in range(n):
        if arr[i][j] == 1:
            bfs(i, j)

ans.sort()

print(len(ans))
for val in ans:
    print(val)

File: algorithm/algorithm_cpp/Algorithm_Session/boj_2667.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>

using namespace std;

constexpr int MAX = 26;
constexpr int SIZE = 4;

static int cnt;
static int rows[] = { -1, 1, 0, 0 };
static int cols[] = { 0, 0, -1, 1 };

int arr[MAX][MAX];
vector<int> rst;

static void bfs(int r, int c, int n) {
	queue<pair<int, int>> q;
	q.push(make_pair(r, c));
	arr[r][c] = 0;

	while (!q.empty()) {
		auto tmp = q.front();
		q.pop();

		cnt++;

		int curR = tmp.first;
		int curC = tmp.second;

		for (int i = 0; i < SIZE; i++) {
			int nextR = rows[i] + curR;
			int nextC = cols[i] + curC;

			if (nextR < 0 || nextC < 0 || nextR >= n || nextC >= n || !arr[nextR][nextC])
				continue;
	
			q.push(make_pair(nextR, nextC));
			arr[nextR][nextC] = 0;
		}
	}

	rst.push_back(cnt);
	cnt = 0;
}

void boj_2667() {
	ios::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);

	int n;
	cin >> n;

	for (int i = 0; i < n; i++)
		for (int j = 0; j < n; j++)
			scanf_s("%1d", &arr[i][j]);

	for (int i = 0; i < n; i++)
		for (int j = 0; j < n; j++)
			if (arr[i][j]) {
				bfs(i, j, n);
				cnt = 0;
			}

	std::sort(rst.begin(), rst.end());

	cout << rst.size() << "\n";
	for (int val : rst)
		cout << val << "\n";
}